### DISCUSSION TITLE
State Parameter for PKCE OAuth flow

### DISCUSSION THREAD

**Discussion by Kysetopher** [supabase/supabase#28520] [Questions]
Hello I am using google sign in. I have successfully setup a PKCE workflow. This flow is working properly and the associated code is listed below. I have also added the state parameter to my initiation. 

It seems that I am being redirected to both supabase and my callback endpoint also included below. however the state parameter is only included in the redirect to supabase. 
Its possible that supabase is handling the state parameter. however I am not sure if I need to verify this on my end as well in the callback by reading the state parameter stored in my cookies. If I do need to verify this im not sure how to read the state parameter when its not being sent to my endpoint. 

login.js endpoint:
```
import { generatePkcePair } from '../../../utils/pkce';
import { setCookie } from 'nookies';
import { ServerClient } from '../../../utils/supabase/server';
import crypto from 'crypto';

export default async function handler(req, res) {
  try {
    const { codeVerifier, codeChallenge } = await generatePkcePair();

    const state = crypto.randomBytes(16).toString('hex'); 

    setCookie({ res }, 'xxxxxxx-auth-token-code-verifier', codeVerifier, {
      path: '/',
      maxAge: 60 * 10,
      secure: true,
      httpOnly: true,
    });

    setCookie({ res }, xxxxxxx-auth-state', state, {
      path: '/',
      maxAge: 60 * 10,
      secure: true,
      httpOnly: true,
    });

    const supabase = ServerClient(req, res);

    const redirectToUrl = process.env.NODE_ENV === 'production'
      ? 'https://www.l33tbox.ai/api/auth/callback'
      : 'http://localhost:3000/api/auth/callback';

      const { data, error } = await supabase.auth.signInWithOAuth({
        provider: 'google',
        options: {
          redirectTo: redirectToUrl,
          scopes: 'email profile',
          codeChallenge,
          codeChallengeMethod: 'S256',
          state: JSON.stringify({ flow_state_id: state }), // Embed state in a JSON object
        },
      });
      
      
      

    if (error) {
      console.error('Supabase OAuth error:', error.message);
      return res.status(500).json({ error: error.message });
    }
    await new Promise(resolve => setTimeout(resolve, 1000));
    console.log('Redirecting to:', data.url);
    return res.redirect(302, data.url);
  } catch (error) {
    console.error('Error during OAuth login:', error.message);
    return res.status(500).json({ error: error.message });
  }
}

```
callback.js endpoint:
```
import { parseCookies, setCookie} from 'nookies';
import { ServerClient } from '../../../utils/supabase/server';

export default async function handler(req, res) {
  try {
    const { code } = req.query;
    const cookies = parseCookies({ req });
    const codeVerifier = cookies['xxxxxxxxxx-auth-token-code-verifier'];

    if (!code || !codeVerifier) {
      throw new Error('Code and code verifier must be non-empty');
    }

    const supabase = ServerClient(req, res);
    const { data, error } = await supabase.auth.exchangeCodeForSession(code, codeVerifier);

    if (error) {
      console.error('Error exchanging code for session:', error);
      throw error;
    }

    const { session } = data;
    if (!session || !session.access_token || !session.refresh_token) {
      throw new Error('Invalid session data');
    }

    setCookie({ res }, 'xxxxxxxx-auth-token', session.access_token, {
      path: '/',
      maxAge: 60 * 60 * 24 * 7, // 1 week
      secure: process.env.NODE_ENV === 'production', // Ensure secure flag is set in production
      httpOnly: true,
      sameSite: 'Strict',
    });

    setCookie({ res }, 'sb-refresh-token', session.refresh_token, {
      path: '/',
      maxAge: 60 * 60 * 24 * 7, // 1 week
      secure: process.env.NODE_ENV === 'production', // Ensure secure flag is set in production
      httpOnly: true,
      sameSite: 'Strict',
    });

    await new Promise(resolve => setTimeout(resolve, 1000));
    console.log('PKCE Oauth flow complete, cookies set, redirecting...');

    res.redirect('/');
  } catch (error) {
    console.error('Error during OAuth redirect handling:', error);
    res.status(500).send(`Error during OAuth redirect handling: ${error.message}`);
  }
}
```

here are my requests that appear in the broswer:

Request URL:
`https://xxxxxxxxxx.supabase.co/auth/v1/callback?state=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJleHAiOjE3MjMyNDA2NDUsInNpdGVfdXJsIjoiaHR0cHM6Ly93d3cubDMzdGJveC5haSIsImlkIjoiMDAwMDAwMDAtMDAwMC0wMDAwLTAwMDAtMDAwMDAwMDAwMDAwIiwiZnVuY3Rpb25faG9va3MiOm51bGwsInByb3ZpZGVyIjoiZ29vZ2xlIiwicmVmZXJyZXIiOiJodHRwOi8vbG9jYWxob3N0OjMwMDAvYXBpL2F1dGgvY2FsbGJhY2siLCJmbG93X3N0YXRlX2lkIjoiYzA5YTZmNWUtZDg1ZS00NjdjLWEzZjctOTc3MzFjZjdlNjU1In0.O86Q7-mkQ_33tjMxArUpOtWy8RHAich9tXm_o71P42w&code=4%2F0AcvDMrDnyQuV6OVtkrGcATftX5Lpb1R_WP8YFfifbzc1Ge9PvT2jSThf-52ATq9F9xQwAQ&scope=email+profile+openid+https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fuserinfo.email+https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fuserinfo.profile&authuser=0&prompt=none
`
Request URL:
`http://localhost:3000/api/auth/callback?code=ebfc2d2b-1318-472d-b4eb-691c3`

**Comment by j4w8n** [comment]
Seems like you're doing a lot of things manually. Have you tried the ssr library? Between that and the supabase-js library, most of this is done for you.

**Comment by Kysetopher** [comment]
Hello,
Thanks for your help. I am using the ssr library, I am not using the
supabase-js library at all. It is very possible that I am not sure how to
use it properly.

Unfortunately I have not found much documentation on what is being handled
for me by the library. I cannot find any documentation that clearly states
that the State Parameter, the Nonce, or the PCKE code challenge is being
verified automatically ( in fact the Google OAuth Documentation
<https://supabase.com/docs/guides/auth/social-login/auth-google?queryGroups=environment&environment=server>
seems
to suggest that I do need to handle at least the nonce in the fashion I am,
however that may be for the other library?). When setting this up it seemed
as though I needed to make a custom storage adapter for the suabase ssr
client because by default I was not able to read any of the cookies that I
had stored data in. this may be messing with my results and may need to be
disabled to use the built in functionality.

However I would love to go through each part of my setup and remove it if
it is safe to do so. I do not feel confident I would be able to do this
without support or more dedicated documentation. In order to ensure that I
am secure I would like to use a PKCE workflow, with a nonce and a state
parameter.

Below are each of my steps, which ones should I remove so that I can let
the library handle them?

*In the initiation:*

1: Generating the PKCE pair

2 : Generating a nonce

3 : Generating a state parameter

4 : Storing the state parameter, nonce, and PKCE code verifier in cookies
to be read later by the callback.

5 : Sending the State Parameter, nonce, and code challenge to google with
signInWithOAuth

*In the Callback:*

6 : reading the state parameter, nonce, and PKCE code verifier from cookies

7 : Reading the Code from the google auth callback url

8: Providing the nonce, the code and the code verifier in exchange for a
session

9: grabbing the session and storing the access token and refresh token in
cookies (this is retrieved later by a validation endpoint that validates
the codes and refreshes the session when the user already logged in)

**Reply by j4w8n** [reply]
I'm not sure of the details on how Supabase implements the PKCE flow - as far as using a nonce and/or state value. However, supabase-js and ssr do pretty much everything else in your list - except you need the callback route, but even in there, you don't do much.

What framework are you using? If Supabase has a guide for it, [here](https://supabase.com/docs/guides/auth/server-side), I'd recommend checking it out.

**Reply by Kysetopher** [reply]
I am using Next.js. I did reference the guide you have linked above when setting up. as well as this one for PKCE:

[PKCE FLOW](https://supabase.com/docs/guides/auth/sessions/pkce-flow)

The only current concern with the flow that I have is that the cookies I am using are exposed to the front end currently, i can see them in the broswer dev tools (I may switch over to server side storage for the authorization tokens to mitigate that). and of course that im not sure how to manage the state parameter. currently the PKCE verification is working as expected and I am able to login securly and recreate accounts when needed. essentially my goal right now is to secure the application against the following types of attacks:

[ ] State Parameter with Oauth : Use a state parameter to prevent CSRF attacks. This needs to be sent with the PCKE code challenge when initiating the handshake with Google. Then when returned the state parameter needs to be checked to ensure it is a match. 

[x] Nonce for  Oauth : This prevents replay attacks, ensuring the token is tied to a specific authentication request and hasn't been tampered with or reused.

Its very possible that supabase is already secure against these attacks but im just not sure to be honest. 

when i was reading the docs youve referenced it actually seemed alot more complex than my current implementation (however I am most likely misunderstanding as there is setup for other things here like email):
A Supabase client reads and sets cookies in order to access and update the user session. 

> Depending on where the client is used, it needs to interact with cookies in a different way:
> 
> getServerSideProps - Runs on the server. Reads cookies from the request, which is passed through from GetServerSidePropsContext.
> getStaticProps - Runs at build time, where there is no user, session, or cookies.
> Component - Runs on the client. Reads cookies from browser storage. Behind the scenes, createBrowserClient reuses the same client instance if called multiple times, so don't worry about deduplicating the client yourself.
> API route - Runs on the server. Reads cookies from the request, which is passed through from NextApiRequest.

In my implementation I am acutally handling everything in next.js endpoints and I do not actually have a client side supabase client. which I felt was more secure. But I could be completely unaware as I am not an expert.

for reference here is my only supabase client which is only ever imported in server side next.js endpoints:

```
import { createServerClient } from '@supabase/ssr';
import { parseCookies, setCookie, destroyCookie } from 'nookies';


export function ServerClient(req, res) {

  return createServerClient(
      process.env.SUPABASE_URL,
      process.env.SUPABASE_KEY,
      {
        auth: {
          detectSessionInUrl: true,
          flowType: 'pkce',
        },
        cookies: {
          getAll: async () => {
            const cookies = parseCookies({ req });
            return Object.entries(cookies).map(([name, value]) => ({ name, value }));
          },
          setAll: async (cookiesToSet) => {
            cookiesToSet.forEach(cookie => {
              setCookie({ res }, cookie.name, cookie.value, {
                path: '/',
                maxAge: 60 * 10, // 10 minutes
                secure: true, // Ensure cookies are sent over HTTPS
                httpOnly: true, // Ensure cookies are not accessible via JavaScript
                ...cookie.options,
              });
            });
          }
        },
      }
    );
  
}
```


this includes my custom storage adapter


so I guess what is not clear to me.: if supabase is handling the PKCE code generation do i just specify that im using PCKE and not provide the code? for example: 

as long as ive specified PKCE here:

```
    auth: {
          detectSessionInUrl: true,
          flowType: 'pkce',
        },
```
It is uncessary to generate the code and upload it here
```
    const { data, error } = await supabase.auth.signInWithOAuth({
      provider: 'google',
      options: {
        redirectTo: redirectToUrl,
        scopes: 'email profile',
        codeChallenge,
        codeChallengeMethod: 'S256',
        state: JSON.stringify({ flow_state_id: state }),
        nonce: hashedNonce,  // Pass the hashed nonce here
      },
    });
```

**Reply by j4w8n** [reply]
The link you shared for the PKCE flow seems to be for a browser implementation, not server side.

The PKCE code challenge is verified against the code-verifier value when you call `supabase.auth.exchangeCodeForSession()`. Technically, it's sent off to the Supabase server for this. [REF](https://github.com/supabase/auth-js/blob/master/src/GoTrueClient.ts#L573)

Your extra `options` don't appear to be passed through on your `signInWithOAuth` call: https://github.com/supabase/auth-js/blob/master/src/lib/types.ts#L584 as I don't believe they are supported.

There's nothing wrong with wanting to do it all yourself, but Supabase tries to make things easier for you by handling the comms to the provider. If your security requirements are beyond what Supabase supports, then you may have to stick with rolling your own. Feel free to open an issue [here](https://github.com/supabase/auth) to ask about how the backend may or may not use a nonce or state value when handling OAuth.

BTW, the `supabase-js` package is a peer dependency of the `ssr` one, so it should be installed and used under-the-hood - even if you're not importing anything from it.

**Reply by Kysetopher** [reply]
Hmm that is interesting. 

I was able to test this. it seems that the only parameter that is needed for the chain is the redirectTo. I commented out all other options and was still able to create new users. 

thank you that is very helpful.

Additionally I was able to inspect the redirect from google to supabase and it is still including the state parameter which means that supabase OAuth likely uses this by default. which is very nice 

thanks so much for your help. ill have to review my implementation and probably will remove my manual settings as it seems they are not actually doing anything at all.
